src/file_pipe_log/pipe.rs

```diff
@@ -295,14 +301,36 @@ impl<F: FileSystem> SinglePipe<F> {
             }
         }
         let start_offset = writer.offset();
-        if let Err(e) = writer.write(bytes.as_bytes(&ctx), self.target_file_size) {
+        let buf_bytes = bytes.as_bytes(&ctx);
+        if let Err(e) = writer.write(buf_bytes, self.target_file_size) {
             if let Err(te) = writer.truncate() {
                 panic!(
                     "error when truncate {} after error: {}, get: {}",
                     seq, e, te
                 );
             }
-            return Err(e);
+            if is_no_space_err(&e) {
+                // To avoid incorrect `rotate_imp` on the same dir when the size of
```

Comment:
I don't understand. If the available space is less than `buf_bytes.len()`, it should panic when actually writing that `buf_bytes`, instead of panicking here too early.

Reply:
In this case, it won't panic but fail on `engine.write(...)` unexpectedly.
Given that we have a two directories setting with `dir` and `spill-dir`. If `dir` has free space for rotating several new log files, but no enough space for dumping the given `buf`. And `spill-dir` has free space for dumping this `buf`.
That is: 
+ `buf_bytes.len()` > `target_file_size` == true;
+ `n * target_file_size` < `dir.spare_space()` && `buf_bytes.len() > dir.spare_space()` && `buf_bytes.len() < spill-dir.spare_space()`. 

Then, with the previous `rotate_imp(self.target_file_size)`,
+ 1. First attempt for writing, failed on `self.write()` and try to rotate.
+ 2. As `target_file_size` < `dir.spare_space()`, `rotate_imp(target_file_size)` successfully, and returned `TryAgain(...)` to trigger the second attempt for write. (Naming this newly rotated file as `rotate-1`.)
+ 3. Secondly, it tried to dump this buf into `rotate-1`. But `self.write()` failed again as the `dir.spare_space()` < `buf_bytes.len()`, and it returned the `NOSPAC` error. Buffer still could not be dumped into current `rotate-1` file. Then, it made the next `rotate_imp()`. And unfortunately, it succeed again because the `dir` still had enough space for rotating a new file. So, it returned `TryAgain()` error again. It would cause the final failure on `engine.write(...)`.
+ 4. The whole write of this LogBatch failed because its trying times on write exceed the trying limitation. However, the `spill-dir` still had enough space for it.

For example, given that, `target_file_size == 1MB`, `dir.spare_space() == 1.5MB`, `spill-dir.spare_space() == 100MB` and `buf_bytes.len() == 3MB`. The above steps would be executed.
```Rust
   Compiling raft-engine v0.3.0 (/nvme1n1/lucasliang/workspace/raft-engine)
    Finished test [unoptimized + debuginfo] target(s) in 13.33s
     Running tests/failpoints/mod.rs (target/debug/deps/failpoints-23610de2336687d7)
thread 'test_engine::test_build_engine_with_spill_dir' panicked at 'called `Result::unwrap()` on an `Err` value: 
TryAgain("Failed to write logbatch, exceed max_retry_count: (2),
err: failed to write 41 file, get IO Error: Custom { kind: StorageFull, error: \"nospace\" } try to flush it to another dir")', 
tests/failpoints/test_engine.rs:37:40
```